microovn/ovn/environment/environment.go

```diff
@@ -104,63 +104,89 @@ func defaultRemoteAddresses(ctx context.Context, s state.State) ([]string, error
+		return "", err
+	}
+
+	// With only a single central node enabled, there are no other cluster members to connect to.
+	// Setting the "initial node" to the address of the only node with the central enabled will
+	// cause the NB/SB databases to be initialized on that node.
+	if len(addrList) == 1 {
+		initialNode = addrList[0]
+	}
+
+	// With more than one node in the central cluster, we set the value of the "initial node" to
+	// point to any non-local node that runs the central service. This will allow the NB/SB databases
+	// to join the existing cluster via the remote node.
+	for _, addr := range addrList {
+		if initialNode == "" && addr != localAddr {
+			initialNode = addr
+			break
+		}
+	}
+
+	if strings.Contains(initialNode, ":") {
+		initialNode = "[" + initialNode + "]"
+	}
+
+	return initialNode, nil
+}
+
+// ConnectionString builds strings that define connection endpoints of OVN
+// cluster services. Its return value can be used to tell processes (i.e. ovn-controller)
+// about the location of OVN central services.
+// Example: "ssl:10.0.0.1:6641,ssl:10.0.0.2:6641,ssl:10.0.0.3:6641"
+func ConnectionString(ctx context.Context, s state.State, port int) (string, error) {
 	// Attempt to get the list of ovn-central nodes from the config.
 	addrList, err := remoteAddressesFromConfig(ctx, s)
 	if err != nil {
-		return "", "", err
+		return "", err
 	}
 
 	// If the option was not set, fall back to using MicroOVN nodes with "central" service enabled
 	if addrList == nil {
 		addrList, err = defaultRemoteAddresses(ctx, s)
 		if err != nil {
-			return "", "", err
+			return "", err
 		}
 	}
 
 	// Transform the list of bare IP addresses into the format understood by the OVN, "<protocol>:<IP>:<PORT>"
 	addresses := make([]string, 0, len(addrList))
-	var initialString string
 	protocol := NetworkProtocol(ctx, s)
-	for i, rawAddress := range addrList {
-		parsedAddr, err := netip.ParseAddr(rawAddress)
-		if err != nil {
-			return "", "", err
-		}
-
-		if i == 0 {
-			initialString = parsedAddr.String()
-			if parsedAddr.Is6() {
-				initialString = "[" + initialString + "]"
-			}
-		}
-
+	for _, nodeAddress := range addrList {
 		addresses = append(
 			addresses,
 			fmt.Sprintf("%s:%s",
 				protocol,
-				net.JoinHostPort(parsedAddr.String(), strconv.Itoa(port)),
+				net.JoinHostPort(nodeAddress, strconv.Itoa(port)),
 			),
 		)
 	}
 
-	return strings.Join(addresses, ","), initialString, nil
+	return strings.Join(addresses, ","), nil
 }
 
 // GenerateEnvironment generates the OVN environment file.
 func GenerateEnvironment(ctx context.Context, s state.State) error {
 	// Get the servers.
-	nbConnect, nbInitial, err := ConnectionString(ctx, s, 6641)
+	nbConnect, err := ConnectionString(ctx, s, 6641)
+	if err != nil {
+		return err
+	}
+
+	sbConnect, err := ConnectionString(ctx, s, 6642)
 	if err != nil {
 		return err
 	}
 
-	sbConnect, sbInitial, err := ConnectionString(ctx, s, 6642)
+	initialNbSb, err := initialNbSbHost(ctx, s)
```

Comment:
When we get here, we have currently called `defaultRemoteAddresses` three times. Said function retrieves and iterates over every member of the cluster, performing at least three string comparisons on every member.

Surely we can do better than that?

Reply:
In the latest iteration, getting the list of "central IPS" was moved to its own function and said list became argument for `ConnectionString` and `initialNbSbHost`, eliminating the need for multiple look-ups/parsing of cluster nodes.